### PR TITLE
Fix: set subscription-type based on message ordering

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Right now, user can define  message processing-guarantees to process/sink messages. however, user can still want to retain message ordering with different processing-types and user should be able to configure FAILOVER subscription type with `ATLEAST_ONCE` option. so, provide option to retain ordering while consuming messages from pulsar-source.

### Modifications

Add `retainOrdering` option in functions-cli like we already have in [cli-Sink](https://github.com/apache/incubator-pulsar/blob/master/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java#L217).
### Result

User can configure message ordering option with different processing types.

### CODE REVIEW

**Comment by rdhabalia** [comment]
rerun cpp tests

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 1]
```diff
@@ -240,6 +240,8 @@ void processArguments() throws Exception {
         protected FunctionConfig.ProcessingGuarantees processingGuarantees;
         @Parameter(names = "--userConfig", description = "User-defined config key/values")
         protected String userConfigString;
+        @Parameter(names = "--retainOrdering", description = "Function consumes and processes messages in order")
```
Comment (srkukarni) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: Not sure if the flag makes sense at this level

Comment (srkukarni) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: Do we think that actually specifying subscription type might be the way to go here. We used to have that flag but removed it.

Comment (rdhabalia) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: > We used to have that flag but removed it.

hmm.. not sure about it but it requires to manage ordering.

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 2]
```diff
@@ -560,24 +565,10 @@ protected FunctionDetails convert(FunctionConfig functionConfig)
                 sourceSpecBuilder.setTopicsPattern(functionConfig.getTopicsPattern());
             }
 
-            // Set subscription type based on processing semantics
-            if (functionConfig.getProcessingGuarantees() != null) {
```
Comment (jerrypeng) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: I don't think we should get rid of this logic.  It is a better user experience if the user just specify a processing guarantee and subscription is automatically set for them.  In general, users will be more familiar with processing guarantee terminology than subscription terminology.  We can have an option for the to override the subscription type from the CLI if the user wants to do that

Comment (rdhabalia) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: >  It is a better user experience if the user just specify a processing guarantee and subscription is automatically set for them

Processing-guarantee generally implies how to ack the message and retain-ordering describes how to consume message. Therefore, if processing-guarantee is `ATMOST_ONCE ` it means, we want to ack the message immediately no-matter what result is and `ATMOST_ONCE` can be used with SHARED or FAILOVER subscription both (means with or without ordering feature). So, generally processing-guarantee and ordering play separate role in terms of consuming and processing the message.

> If a user specifies a processing guarantee of effectively once, the subscription by default should be Failover

hmm.. yes, that make sense. I will make that change.

Comment (rdhabalia) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: @jerrypeng  fixed it.

**Comment by jerrypeng** [comment]
@rdhabalia @merlimat @sijie @srkukarni 
I don't believe we have to correct mechanism for setting the subscription type in this PR as well as in CmdSinks base off of the retainOrder flag.  

For functions, sources, and sinks, if the processing guarantee is set to effectively once, it is implied that order should be maintained and the subscription type to be failover.  Only for at-most-once and at-once processing guarantees should we allow the user the change/choose between failover and shared.  I think we should bring back the subscriptionType flag for users to be able to override the subscription type when in at-most-once or at-least-once mode.

**Comment by srkukarni** [comment]
+1 to what @jerrypeng said

**Comment by sijie** [comment]
looks like --subscription-type to allow people override the subscription for at-least-once and at-most-once processes is a better approach. 

also let's also improve the command line description to make things clearer

**Comment by rdhabalia** [comment]
> looks like --subscription-type to allow people override the subscription for at-least-once and at-most-once processes is a better approach.

in #2200 [first commit](https://github.com/apache/incubator-pulsar/pull/2200/commits/6757851bf86b1b8daa0137463ecbf44875075af8) had change to add `--subscription-type` but based on feedback we have added flag for `retain ordering`.

> For functions, sources, and sinks, if the processing guarantee is set to effectively once, it is implied that order should be maintained and the subscription type to be failover. 

Isn't this PR address the same thing. if `guarantee is set to effectively once` then sub type will be set as FAILOVER else sub-type will be configured as per ordering-flag.  is there anything we are missing here?

**Comment by rdhabalia** [comment]
retest this please

**Pull request merged**